Classes/Event/EventWithMetadata.php

```diff
@@ -36,8 +36,13 @@
      */
     public function __construct(EventInterface $event, array $metadata)
     {
-        $this->event = $event;
-        $this->metadata = ($event instanceof EventWithMetadata) ? Arrays::arrayMergeRecursiveOverrule($event->getMetadata(), $metadata) : $metadata;
+        if ($event instanceof EventWithMetadata) {
+            $this->event = $event->getEvent();
+            $this->metadata = Arrays::arrayMergeRecursiveOverrule($event->getMetadata(), $metadata);
+        } else {
+            $this->event = $event;
+            $this->metadata = $metadata;
+        }
```

Comment:
I sneaked in this improvement as follow-up to #156, so that `EventWithMetadata` can actually be nested